fs: readdir optionally returning type information

bengl opened the pull request:
readdir and readdirSync now have a "withTypes" option, which, when enabled,
provides an array of DirectoryEntry objects, similar to Stats bjects,
which have the filename and the type information.

Fixes: https://github.com/nodejs/node/issues/15699

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

mscdex commented on src/node_file.cc:
```diff
@@ -598,6 +598,65 @@ void AfterScanDir(uv_fs_t* req) {
+        name_idx = 0;
+      }
+
+      name_argv[name_idx++] = Integer::New(env->isolate(), ent.type);
```
I know at one time V8 would optimize the case for arrays where each element was of the same type. Perhaps this might perform better if we had a separate array for the type values? That would also allow us to avoid having to `i \ 2` inside the JS loop, which is slower than just referencing `i`.

vsemozhetbyt commented on doc/api/fs.md:
```diff
@@ -283,6 +283,87 @@ synchronous use libuv's threadpool, which can have surprising and negative
+option set to `true`, the resulting array is filled with `fs.DirectoryEntry`
+objects, rather than strings or `Buffers`.
+
+### dirent.name
```
1. Missing YAML?
2. It seems this property should be the last one, ABC-wise.

vsemozhetbyt commented on doc/api/fs.md:
```diff
@@ -283,6 +283,87 @@ synchronous use libuv's threadpool, which can have surprising and negative
+
+* Returns: {boolean}
+
+Returns `true` if the `fs.DirectoryEntry` object describes a file system directory.
```
Exceeds 80 characters length.

vsemozhetbyt commented on doc/api/fs.md:
```diff
@@ -283,6 +283,87 @@ synchronous use libuv's threadpool, which can have surprising and negative
+
+* Returns: {boolean}
+
+Returns `true` if the `fs.DirectoryEntry` object describes a first-in-first-out (FIFO)
```
Exceeds 80 characters length.

bengl commented:
@addaleax @vsemozhetbyt I added a squash commit which I think addresses your comments. PTAL.

@mscdex What would be better as a return value: an array of arrays like `[[name1, name2, ...], [type1, type2, ...]]` or an object containing both arrays? Or an array of `{ name, type }` objects?

bengl commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16072/

devsnek commented:
why don't we just add scandir and scandirSync so we don't have to deal with polymorphic return types and weird options

also +1 to `{ name, type }`, the whole stats object thing is terrible

mscdex commented:
> What would be better as a return value: an array of arrays like [[name1, name2, ...], [type1, type2, ...]] or an object containing both arrays? Or an array of { name, type } objects?

I would guess it's still faster to create objects in JS land, but what I was suggesting is to pass back two separate arrays to `req.oncomplete()`. You could even reuse one of the arrays then:

```js
req.oncomplete = (err, names, fileTypes) => {
  if (err) {
    callback(err);
    return;
  }
  const len = names.length;
  for (var i = 0; i < len; ++i)
    names[i] = new DirectoryEntry(names[i], fileTypes[i]);
  callback(null, names);
};
```

I'm not sure what V8 does optimization-wise when reusing the array like that, if it re-optimizes after the loop finishes or what, but that's the general idea I had in mind.

ronkorving commented:
I'm definitely with @devsnek on making these separate API. Having one API return completely different things based on options is .. a bit too much. Different input format causing a different output format... At that point, I think it has just become a different function.

bengl commented:
@mscdex Right, that would make sense for the async case, but for the sync case, since a single value needs to be returned from the binding function, something holding both arrays is needed.

@devsnek 
> why don't we just add scandir and scandirSync so we don't have to deal with polymorphic return types and weird options

I had originally planned on doing it that way (as suggested in the linked issue), but most functions on `fs`, including `readdir`/`readdirSync` define themselves in terms of UNIX man pages. In there, [`scandir`](http://man7.org/linux/man-pages/man3/scandir.3.html) is defined as "scan a directory for matching entries", wheras what's happening here is far closer to [`readdir`](http://man7.org/linux/man-pages/man3/readdir.3.html). I'd rather reserve `scandir` for when something that closer matches it is implemented. (Note: I'm not necessarily a big fan of having such a POSIX-centric `fs` module, but it's what we have, and my intention was to not stray too far from it.)

> also +1 to { name, type }, the whole stats object thing is terrible

I did the `Stats`-esque `DirectoryEntry` object deliberately to be consistent with the `stat` functions, since any use case of this new feature was previously worked around by calling `stat` on each entry returned by `readdir`/`readdirSync`. Also, the type is a number, and I didn't think it was best to either convert it on the fly to something readable or force the comparison onto the user.

All that being said: If there's consensus around `scandir`/`scandirSync`, bare-object-dirents, or any other significant changes to the API I've presented here, I'm happy to make those changes.

mscdex commented:
> but for the sync case, since a single value needs to be returned from the binding function, something holding both arrays is needed.

A 2-D array would probably work for that case.

Perhaps an even better solution that would work for both scenarios would be to create the zero-length arrays in JS land and pass them (via the context object) to and fill them in in C++ land.

bengl commented:
@mscdex Since the async version seems to be all set up to handle both callbacks and promises in C++, it seemed easier to just have a 2-D array in all cases, which I've done in the latest squash commit.

---

And another CI run...

CI: https://ci.nodejs.org/job/node-test-pull-request/16073/

BridgeAR commented on lib/fs.js:
```diff
@@ -763,9 +780,19 @@ function readdirSync(path, options) {
-  return result;
+  if (!options.withFileTypes) {
+    return result;
+  } else {
```
Super tiny nit: the else could be removed to reduce indentation. The same in other places.

bengl commented:
I'm not sure what the AIX and SmartOS errors are all about. It seems like `isFile` isn't working on those systems? That's pretty weird. Maybe the values for the types are not in the set defined by libuv?

bengl commented:
Here's another CI run, hopefully fixing AIX and SmartOS...

CI: https://ci.nodejs.org/job/node-test-pull-request/16088/

Trott commented:
Resume Build: https://ci.nodejs.org/job/node-test-pull-request/16109/

bengl commented:
Since this introduces a binding-layer breaking change....

CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1480/

bengl commented:
Rebased to deal with `FSReqWrap` -> `FSReqCallback` from @maclover7.

New CI: https://ci.nodejs.org/job/node-test-pull-request/16131/

bengl commented:
Post-rebase-fix CI: https://ci.nodejs.org/job/node-test-pull-request/16134/

addaleax commented on doc/api/fs.md:
```diff
@@ -283,6 +283,97 @@ synchronous use libuv's threadpool, which can have surprising and negative
+Note that on some systems, getting file types at the same time as `readdir` is
+not supported. On those systems, only `dirent.isUnknown()` will return true, and
+all other `DirectoryEntry` methods will return false. To get file types on those
+systems, use [`fs.stat()`][].
```
I’m wondering whether it might make sense to use `fs.stat()` in those cases in Node.js. The user wouldn’t be requesting `withFileTypes` if they aren’t going to use it, right? I know it sounds like a lot of overhead, but moving that overhead to userland doesn’t really buy us that much, does it?

bengl replied:
Good call I'll implement this when I have some time in the next week or so.

bengl replied:
The latest squash commit I pushed does this. PTAL.

addaleax replied:
Does that mean that this note can be removed now? :)

silverwind commented:
I don't think it fixes https://github.com/nodejs/node/issues/15699 because for me, the draw to native `scandir` is performance, so we certainly want to reserve that name for a future implementation of it (current effort at https://github.com/libuv/libuv/pull/416).

silverwind commented:
Benchmarking results look good. I tested in Linux with dropping caches before each run. The script is recursively walking the node repository, sorting found files/directories into arrays.

`readdir` and `stat` sync: 893 ms
`readdir` and `stat` async: 1158 ms
`readdir` with `withTypes` option sync: 424 ms
`readdir` with `withTypes` option async: 349 ms

Quite surprised that async was faster there.

For my specific use case, I'd also need the `stat` object, so I can't really benefit from this new option. Python's [scandir](https://docs.python.org/3/library/os.html#os.DirEntry) has a  [`stat`](https://docs.python.org/3/library/os.html#os.DirEntry.stat) method on each `DirEntry` and the docs say it would work without an additional system call on non-symlinks on Windows, so this could be a possible future addition, if libuv supports it.

mscdex commented on src/node_file.cc:
```diff
@@ -551,51 +551,139 @@ void AfterScanDir(uv_fs_t* req) {
+    }
   }
+
+  Local<Array> result = Array::New(env->isolate(), 2);
```
Perhaps keeping a local constant array in JS, which is passed to the C++ layer, can be reused instead of creating a new array of length 2 every time? Just be sure to empty out the contents afterwards to allow GC.

bengl replied:
Hmm, I'm not too sure about the mechanics of this, due to the async version. Would stashing the array on the `FSReqBase` be the thing to do here? Or in the node environment (and then only passing it in once)?

mscdex replied:
There would be no problem for async since setting the two array elements and processing them would happen on the main thread and would happen at the same time.

As far as how to allocate the array only once, you could do it any which way really. I don't really have a preference.

addaleax replied:
I wouldn’t expect all too much overhead from creating a new array in C++ every time, so maybe we can leave it at this…?

Something where we probably can save time is using a typed array for the second entry, instead of a plain array of integers.

mscdex replied:
Last I checked creating arrays or objects in C++ was significantly slower. Besides I would imagine having less pressure on the GC would be a good thing?

addaleax replied:
It is going to be faster in JS, yes. I just don’t think it’s going to have an effect in the big picture, when compared to all the other work this code does, and I’m not a fan of optimizations like these when we haven’t measured their impact as part of their actual use case…

bengl replied:
@mscdex @addaleax Maybe the thing to do then, is punt on this for now, and do an optimization PR later after seeing if it turns out to be significant?

devsnek commented on lib/internal/fs/promises.js:
```diff
@@ -37,10 +38,13 @@ const {
 const kHandle = Symbol('handle');
 const { kUsePromises } = binding;
 
+const getDirectoryEntriesPromise = promisify(getDirectoryEntries);
```
can you just abstract the native call? promisify isn't ideal...

bengl replied:
Can you clarify? Which native call? The only call `getDirectoryEntries` makes into binding code is when there's unknown types and so it has to call `stat`, which is why it's asynchronous (although it's synchronous using `statSync` when used by `readdirSync`), and used here as promisified.

I could make a promise-specific version of `getDirectoryEntries`, using the promise version of stat under the hood, but that seemed like overkill. (Maybe it's not?)

bengl commented:
@silverwind 

> I think it this point, I'm +1 at just calling it `scandir`.

I outlined here <https://github.com/nodejs/node/pull/22020#issuecomment-408761439> why I chose to keep it as `readdir` with an option. Is that ok? Or do you want to block on it?

jdalton commented on lib/internal/fs/utils.js:
```diff
@@ -31,24 +33,139 @@ const {
+function lazyLoadFs() {
+  if (!fs) {
+    fs = require('fs');
+  }
```
Can we verify that `fs` isn't already loaded before this.

bengl replied:
`internal/fs/utils` is required by `fs`, so the dependency is circular. That's why lazily loading it is necessary.

silverwind commented:
> I outlined here #22020 (comment) why I chose to keep it as readdir with an option. Is that ok

Fine with me, thought I don't like this discrepancy with Python it creates.

@bengl two questions related to features python's `os.scandir` has:

1) Would it be costly to add a `path` property with the full path to each entry? I assume it would be useful in many cases and if we already have it available, I'd say provide it to the user.
2) Would you know if syscall-less stats on Windows possible (see [here](https://docs.python.org/3/library/os.html#os.DirEntry.stat))?

jdalton commented on doc/api/fs.md:
```diff
@@ -283,6 +283,97 @@ synchronous use libuv's threadpool, which can have surprising and negative
+`fs.DirectoryEntry` objects, rather than strings or `Buffers`.
+
+Note that on some systems, getting file types at the same time as `readdir` is
+not supported. On those systems, only `dirent.isUnknown()` will return true, and
```
The mix of `DirectoryEntry` and `dirent` got me.
If we call it `Stat` and not `Status` maybe `DirectoryEntry` should be called `Dirent`.

bengl replied:
@jdalton I'm not sure we gain much from brevity, since users don't ever even need to use the class name in order to use this feature. What if I just change `dirent` to `directoryentry`?

bengl commented:
@silverwind 

> 1. Would it be costly to add a path property with the full path to each entry? I assume it would be useful in many cases and if we already have it available, I'd say provide it to the user.

Concatenating strings per entry does some like something that could add up, cost-wise, and doesn't really seem in scope for this PR, whose purposes is just to get types at the same time as entry names. Maybe this can be done as a future addition? Maybe it should be optional?

> 2. Would you know if syscall-less stats on Windows are possible?

If I'm reading the python doc you linked correctly, then that may also be possible, yes. That being said, I don't know much about Windows APIs.

bengl commented:
@jdalton I renamed `DirectoryEntry` to `Dirent`.

apapirovski commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16297/

mrmlnc commented on lib/internal/fs/utils.js:
```diff
@@ -31,24 +33,139 @@ const {
+        const name = names[i];
+        const idx = i;
+        toFinish++;
+        lazyLoadFs().stat(pathModule.resolve(path, name), (err, stats) => {
```
Are we sure we want to have this behavior?

At a minimum, for me, DirEntry is associated with entry information, including information that there is no type information for a given entry.

It's confusing.

bengl replied:
This behavior was specifically asked for by @addaleax in this comment: https://github.com/nodejs/node/pull/22020#discussion_r207039181

`isUnknown()` will still return true for entries whose type information was retrieved via `fs.stat`, so for the use case you're describing, information is not lost.

mrmlnc replied:
So, in this case, the `isUnknown` method is not described in the documentation (`doc/api/fs.md` `Dirent` class).

But, unfortunately, I still don't understand why the low-level tool has an implicit action.

For example, I intended to make additional logic for unknown-entries in the package:

> Read the directory. Get a set of entries. If the user wants to see `fs.Stats` for entries, then we get it.

In the current situation we will receive the `fs.Stats` twice (no big deal) if the system doesn't give us types to entries. I understand that this is a degenerate example.

P.S.: That's interesting, can the same system provide a type for some entries, and for some not.

❤️ In any case, thank you for your work!

bengl replied:
> So, in this case, the `isUnknown` method is not described in the documentation (`doc/api/fs.md` `Dirent` class).

That's an mistake on my part. Whoops! I'll add it in.

> But, unfortunately, I still don't understand why the low-level tool has an implicit action.

The intent (I think, based on the comment by @addaleax) is to actually have directory entry type information in as many cases as possible, and not depend on the user code to perform an extra action to get the information they asked for in the first place in the event of an easily recoverable bad case (i.e. `isUnknown`). This seems reasonable to me.

> In the current situation we will receive the fs.Stats twice (no big deal) if the system doesn't give us types to entries. I understand that this is a degenerate example.

Perhaps a future optimization PR could cache the results of the first `stat` call? (Maybe clearing that cache on the next tick?)

> P.S.: That's interesting, can the same system provide a type for some entries, and for some not.

Excellent question! To be honest, I'm not really sure what the exhaustive list of situations is, in which `isUnknown` returns true. I do know that it seemed to show up reliably in CI on SmartOS and AIX for entries that I knew were ordinary files, and `fs.stat` works just fine on them. I'm basically just going with what libuv reports.

bengl replied:
@mrmlnc Sorry I looked over my code again and it seems that my intention was actually to *remove* the `isUnknown` method, because it's now unnecessary, since the underlying `fs.stat` call will make the type always knowable. I think that's the best path to take here, so I'll go ahead and remove `isUnknown` from the implementation and the test file.

That being said, if you do find yourself in a situation where you want to know whether `fs.stat` has been called internally, note that the `Dirent` instances in that case will actually be instances of `DirentFromStats`, an undocumented subclass I used in order to make it easy to have the methods from `Stats` fill in the missing data. So you can always detect that if you really need that information.

Sorry for the confusion! It's getting late here (UTC-7), haha.

bengl commented:
It looks like (if I'm not mistaken) this is in a state of approval, so if no one objects, and CI and CITGM seem healthy, I'll land this on Monday or Tuesday next week.

Also I'll update the commit message in reference to the new option name, and change `Fixes` to `Ref`, since it seems related but not totally solving for that use case.

bengl commented:
Resume Build: https://ci.nodejs.org/job/node-test-pull-request/16397/

mscdex commented:
@bengl This is unrelated and I know it's only temporary until everything gets squashed, but in the future could you please avoid including my github handle mention in commit messages? I get a notification *every* time a (force) push is made with the mention in the commit message. I understand the desire to include attribution, but for me personally I do not mind not having attribution for suggested changes.

Thanks.

bengl commented:
@mscdex Whoops! I wasn't aware Github did that for commit messages. In the future I'll avoid that. Sorry for all the noise! FWIW I think that was probably the last push before squashing and landing, so that should be the end of it.

silverwind commented:
If possible, I'd like to see this backported to 8.x.

ajafff commented on lib/internal/fs/utils.js:
```diff
@@ -31,24 +33,135 @@ const {
+      const type = types[i];
+      if (type === UV_DIRENT_UNKNOWN) {
+        const name = names[i];
+        const stats = lazyLoadFs().statSync(pathModule.resolve(path, name));
```
Maybe that was asked before (I didn't read through all comments): why does this use `resolve` instead of `join`? At this point `name` is not expected to be an absolute path.

ajafff replied:
Why does this use `statSync` instead of `lstatSync`? This makes it rather inconsistent because in one case a Dirent may be reported as symbolic link and in case `readdir` gives `type === UV_DIRENT_UNKNOWN` symlinks are resolved and the type of the symlinked file is used instead.
